File: 0055-jump-game/0055-jump-game.java
class Solution {
    public boolean canJump(int[] nums) {
        //at each step take whats the max jump u can take 
        int curJumps=0;
        if(nums.length==1) return true;
        if(nums[0]==0) return false;
        for(int i=0;i<nums.length-1;i++){
            curJumps--;
            if(nums[i] > curJumps){
                curJumps=nums[i];
            }
            if(curJumps==0) return false;
        }
        if(curJumps>0) return true;
        return false;
    }
}

File: 0055-jump-game/0055-jump-game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        int  n = nums.size();
        int curFar=0,curEnd =0;
        for(int i=0;i<n-1;i++){
            curFar = max(curFar,i+nums[i]);
                if(i==curEnd){
                    curEnd=curFar;
                }
        }
        if(curEnd<n-1){
            return false;
        }
        return true;
    }
};